pkg/controller/controller_sync_task.go

```diff
@@ -156,7 +167,78 @@ func (st *syncTask) evalSpec(res *smith_v1.Resource) (*unstructured.Unstructured
 			BlockOwnerDeletion: &trueRef,
 		})
 	}
-	spec.SetOwnerReferences(refs)
+	obj.SetOwnerReferences(refs)
+
+	return obj, nil
+}
+
+// evalPluginSpec evaluates the plugin resource specification and returns the result.
+func (st *syncTask) evalPluginSpec(res *smith_v1.Resource) (*unstructured.Unstructured, error) {
+	plugin, ok := st.plugins[res.PluginName]
+	if !ok {
+		return nil, errors.Errorf("no such plugin %q", res.PluginName)
+	}
+	dependencies := make(map[smith_v1.ResourceName]smithPlugin.Dependency, len(res.DependsOn))
+	for _, name := range res.DependsOn {
+		var dependency smithPlugin.Dependency
+		unstructuredActual := st.readyResources[name]
+		gvk := unstructuredActual.GroupVersionKind()
+		actual, err := st.scheme.New(gvk)
+		if err != nil {
+			return nil, err
+		}
+		if err = unstructured_conversion.DefaultConverter.FromUnstructured(unstructuredActual.Object, actual); err != nil {
+			return nil, err
+		}
+		dependency.Actual = actual
+		switch obj := actual.(type) {
+		case *sc_v1b1.ServiceBinding:
+			secret, exists, err := st.store.Get(core_v1.SchemeGroupVersion.WithKind("Secret"), obj.Namespace, obj.Spec.SecretName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error finding output Secret for ServiceBinding %q", obj.Name)
+			}
+			if !exists {
+				return nil, errors.Errorf("cannot find output Secret for ServiceBinding %q", obj.Name)
+			}
+			dependency.Outputs = append(dependency.Outputs, secret)
+
+			serviceInstance, exists, err := st.store.Get(sc_v1b1.SchemeGroupVersion.WithKind("ServiceInstance"), obj.Namespace, obj.Spec.ServiceInstanceRef.Name)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error finding ServiceInstance for ServiceBinding %q", obj.Name)
+			}
+			if !exists {
+				return nil, errors.Errorf("cannot find ServiceInstance for ServiceBinding %q", obj.Name)
+			}
+			dependency.Auxiliary = append(dependency.Auxiliary, serviceInstance)
+		}
+		dependencies[name] = dependency
+	}
+	log.Printf("Plugin %s dependencies: %+v", res.PluginName, dependencies)
+	result, err := plugin(*res, dependencies)
+	if err != nil {
+		return nil, err
+	}
+	// TODO validate pluginSpec GVK/name against result.Object.Spec
+	// and validate type of object against schema if possible
+
+	log.Printf("Plugin %s result: %+v", res.PluginName, result.Object)
+	return util.RuntimeToUnstructured(result.Object)
+}
+
+// evalNormalSpec evaluates the regular resource specification and returns the result.
+func (st *syncTask) evalNormalSpec(res *smith_v1.Resource) (*unstructured.Unstructured, error) {
+	// 0. Convert Spec to Unstructured
```

Comment:
nit: you count steps from 1 in `evalSpec` and from 0 here.
make consistent, or remove ordinary numbers completely.